src/charm.py

```diff
@@ -31,12 +31,61 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
```

Comment:
Hi @jnsgruk, good question! @marceloneppel and I thought that it would make things cleaner/easier to understand. If you think it makes things confusing let me know and I can revert :)

Reply:
Nevermind, I will opt for your suggestion below